File: crons-realm/src/main/java/com/raxdenstudios/cron/model/Cron.java
package com.raxdenstudios.cron.model;

import android.app.AlarmManager;

import io.realm.RealmModel;
import io.realm.annotations.PrimaryKey;
import io.realm.annotations.RealmClass;

@RealmClass
public class Cron implements RealmModel {

    @PrimaryKey
    private long id;
    private int type;
    private long triggerAtTime;
    private long interval;
    private boolean status;
    private String data;

    public Cron() {

    }

    private Cron(Builder builder) {
        this.id = builder.id;
        this.triggerAtTime = builder.triggerAtTime;
        this.interval = builder.interval;
        this.type = builder.type;
        this.status = builder.status;
        this.data = builder.data;
    }


    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    public int getType() {
        return type;
    }

    public void setType(int type) {
        this.type = type;
    }

    public long getTriggerAtTime() {
        return triggerAtTime;
    }

    public void setTriggerAtTime(long triggerAtTime) {
        this.triggerAtTime = triggerAtTime;
    }

    public long getInterval() {
        return interval;
    }

    public void setInterval(long interval) {
        this.interval = interval;
    }

    public boolean isStatus() {
        return status;
    }

    public void setStatus(boolean status) {
        this.status = status;
    }

    public String getData() {
        return data;
    }

    public void setData(String data) {
        this.data = data;
    }

    public static class Builder {

        private long id;
        private int type;
        private long triggerAtTime;
        private long interval;
        private boolean status;
        private String data;

        public Builder() {
            this(0);
        }

        public Builder(long id) {
            this(id, 0, 0);
        }

        public Builder(long id, long triggerAtTime, long interval) {
            this(id, triggerAtTime, interval, AlarmManager.RTC_WAKEUP);
        }

        public Builder(long id, long triggerAtTime, long interval, int type) {
            this(id, triggerAtTime, interval, type, true);
        }

        public Builder(long id, long triggerAtTime, long interval, int type, boolean status) {
            this.id = id;
            this.triggerAtTime = triggerAtTime;
            this.interval = interval;
            this.type = type;
            this.status = status;
        }

        public Builder id(long id) {
            this.id = id;
            return this;
        }

        public Builder type(int type) {
            this.type = type;
            return this;
        }

        public Builder triggerAtTime(long triggerAtTime) {
            this.triggerAtTime = triggerAtTime;
            return this;
        }

        public Builder interval(long interval) {
            this.interval = interval;
            return this;
        }

        public Builder status(boolean status) {
            this.status = status;
            return this;
        }

        public Builder data(String data) {
            this.data = data;
            return this;
        }

        public Cron create() {
            return new Cron(this);
        }

    }

}

File: crons/src/main/java/com/raxdenstudios/cron/model/Cron.kt
package com.raxdenstudios.cron.model

import android.app.AlarmManager

data class Cron(val builder: Builder) {

    val id: Long
    val type: Int
    val triggerAtTime: Long
    val interval: Long
    val status: Boolean

    init {
        id = builder.id
        type = builder.type
        triggerAtTime = builder.triggerAtTime
        interval = builder.interval
        status = builder.status
    }

    class Builder(internal var id: Long = 0) {

        internal var triggerAtTime: Long = 0
        internal var interval: Long = 0
        internal var type: Int = AlarmManager.RTC_WAKEUP
        internal var status: Boolean = true
        internal var data: String? = null

        fun id(id: Long): Builder {
            this.id = id
            return this
        }

        fun type(type: Int): Builder {
            this.type = type
            return this
        }

        fun triggerAtTime(triggerAtTime: Long): Builder {
            this.triggerAtTime = triggerAtTime
            return this
        }

        fun interval(interval: Long): Builder {
            this.interval = interval
            return this
        }

        fun status(status: Boolean): Builder {
            this.status = status
            return this
        }

        fun data(data: String): Builder {
            this.data = data
            return this
        }

        fun create(): Cron {
            return Cron(this)
        }

    }

}